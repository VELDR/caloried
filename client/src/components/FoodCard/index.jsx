import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { createStructuredSelector } from 'reselect';
import { injectIntl } from 'react-intl';
import { Skeleton, Tooltip } from '@mui/material';
import proteinIcon from '@static/images/protein.png';
import fatIcon from '@static/images/fat.png';
import carbsIcon from '@static/images/carbs.png';
import caloriesIcon from '@static/images/calories.svg';
import { ChevronRight } from '@mui/icons-material';
import { selectLoading } from '@containers/App/selectors';

import classes from './style.module.scss';

const FoodCard = ({ food, onClick, intl: { formatMessage }, loading }) => (
  <div className={classes.container} onClick={onClick}>
    <div className={classes.food}>
      <div className={classes.imageWrap}>
        {loading ? (
          <Skeleton variant="rounded" className={classes.skeletonImage} />
        ) : (
          <>
            <img src={food?.image} alt={food?.foodName} className={classes.image} />

            <span
              className={
                food?.type === 'common' ? `${classes.type} ${classes.common}` : `${classes.type} ${classes.branded}`
              }
            >
              {food?.type}
            </span>
          </>
        )}
      </div>
      <div className={classes.food__content}>
        <Tooltip title={food?.foodName}>
          <div className={classes.name}>{loading ? <Skeleton /> : food?.foodName}</div>
        </Tooltip>
        <div className={classes.serving}>
          {loading ? (
            <Skeleton width={150} />
          ) : (
            <>
              <Tooltip title={`${food?.servingQty} ${food?.servingUnit}`}>
                <span className={classes.serving__unit}>
                  {food?.servingQty} {food?.servingUnit}
                </span>
              </Tooltip>
              {food?.servingWeight && <span className={classes.serving__weight}>â€¢ {food?.servingWeight} g</span>}
            </>
          )}
        </div>
        {loading ? (
          <Skeleton className={classes.skeletonNutrient} />
        ) : (
          <div className={classes.nutrient}>
            <div className={classes.nutrient__item}>
              <img src={caloriesIcon} alt="" className={classes.icon} />
              {food?.nutrients?.calories?.value}
            </div>
            <div className={classes.nutrient__item}>
              <img src={proteinIcon} alt="" className={classes.icon} />
              {food?.nutrients?.protein?.value}
            </div>
            <div className={classes.nutrient__item}>
              <img src={carbsIcon} alt="" className={classes.icon} />
              {food?.nutrients?.carbs?.value}
            </div>
            <div className={classes.nutrient__item}>
              <img src={fatIcon} alt="" className={classes.icon} />
              {food?.nutrients?.fat?.value}
            </div>
          </div>
        )}
      </div>
    </div>

    <div className={classes.right}>
      <div className={classes.nutrientBig}>
        {loading ? (
          <Skeleton width={250} height={50} />
        ) : (
          <>
            <Tooltip title={formatMessage({ id: 'app_protein' })}>
              <div className={classes.nutrientBig__item}>
                <img src={proteinIcon} alt="" className={classes.icon} />
                {food?.nutrients?.protein?.value} g
              </div>
            </Tooltip>
            <Tooltip title={formatMessage({ id: 'app_carbs' })}>
              <div className={classes.nutrientBig__item}>
                <img src={carbsIcon} alt="" className={classes.icon} />
                {food?.nutrients?.carbs?.value} g
              </div>
            </Tooltip>
            <Tooltip title={formatMessage({ id: 'app_fat' })}>
              <div className={classes.nutrientBig__item}>
                <img src={fatIcon} alt="" className={classes.icon} />
                {food?.nutrients?.fat?.value} g
              </div>
            </Tooltip>
          </>
        )}
      </div>
      <div className={classes.rightMost}>
        {loading ? (
          <Skeleton className={classes.skeletonCalories} />
        ) : (
          <div className={classes.calories}>
            <span className={classes.calories__value}>{food?.nutrients?.calories?.value}</span> kcal
          </div>
        )}
        <ChevronRight className={classes.chevronRight} />
      </div>
    </div>
  </div>
);

FoodCard.propTypes = {
  food: PropTypes.object,
  onClick: PropTypes.func,
  intl: PropTypes.object,
  loading: PropTypes.bool,
};

const mapStateToProps = createStructuredSelector({
  loading: selectLoading,
});

export default injectIntl(connect(mapStateToProps)(FoodCard));
