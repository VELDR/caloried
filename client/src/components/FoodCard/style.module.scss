@import '@styles/base';

.container {
  cursor: pointer;
  border-bottom: 1px solid var(--color-text-secondary);
  padding: 0.5rem 0.25rem;
  @include flex-items-center(3rem);
  justify-content: space-between;
  &:hover {
    opacity: 0.75;
  }
  &:hover .chevronRight {
    color: var(--color-primary);
  }
  .food {
    @include flex-items-center(0.5rem);
    .imageWrap {
      @include box(4rem);
      @include flex-center;
      position: relative;

      .image {
        max-width: 4rem;
        max-height: 4rem;
        border-radius: 0.5rem;
        object-fit: cover;
      }
      .type {
        position: absolute;
        bottom: 0;
        width: 100%;
        text-align: center;
        text-transform: capitalize;
        border-bottom-left-radius: 0.5rem;
        border-bottom-right-radius: 0.5rem;
        font-size: $font-size-tiny;
        color: var(--color-bg-primary);
      }
      .common {
        background-color: var(--color-primary);
      }
      .branded {
        background-color: var(--color-accent);
      }
    }
    &__content {
      @include flex-column;

      .name {
        text-transform: capitalize;
        font-weight: 700;
        @include truncate(13rem);

        @include apply-at-sm-size {
          @include truncate(23rem);
        }

        @include apply-at-md-size {
          @include truncate(11rem);
        }
      }
      .serving {
        @include flex-row(0.25rem);
        font-size: $font-size-regular;
        color: var(--color-text-secondary);

        &__unit {
          @include truncate(6rem);
          @include apply-at-sm-size {
            @include truncate(12rem);
          }
          @include apply-at-md-size {
            @include truncate(7rem);
          }
        }
      }
    }
    .nutrient {
      @include flex-items-center(0.5rem);
      @include apply-at-md-size {
        display: none;
      }
      &__item {
        @include flex-items-center(0.25rem);
        font-size: $font-size-smaller;
      }
      .icon {
        @include box(0.75rem);
      }
    }
  }

  .right {
    @include flex-items-center(2rem);
    .nutrientBig {
      display: none;
      @include apply-at-md-size {
        display: block;
        @include flex-items-center(1rem);
        font-size: $font-size-small;
      }
      &__item {
        @include flex-items-center(0.25rem);
        .icon {
          @include box(1.5rem);
        }
      }
    }
  }
  .leftMost {
    @include flex-items-center;
    .calories {
      font-size: $font-size-small;
      display: none;
      @include apply-at-md-size {
        display: block;
      }
      &__value {
        font-size: $font-size-large;
      }
    }
  }
}
