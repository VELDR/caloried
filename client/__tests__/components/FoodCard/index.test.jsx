import FoodCard from '@components/FoodCard';
import { fireEvent, render } from '@utils/testHelper';

let wrapper;

jest.mock('@static/images/protein.png', () => 'path/to/protein.png');
jest.mock('@static/images/carbs.png', () => 'path/to/carbs.png');
jest.mock('@static/images/fat.png', () => 'path/to/fat.png');

const mockFood = {
  foodName: 'Test Food',
  image: 'test-image.jpg',
  type: 'common',
  servingQty: 1,
  servingUnit: 'serving',
  servingWeight: 100,
  nutrients: {
    calories: { value: 200 },
    protein: { value: 10 },
    carbs: { value: 20 },
    fat: { value: 5 },
  },
};

const mockProps = {
  food: mockFood,
  onClick: jest.fn(),
  intl: {
    formatMessage: jest.fn(),
  },
  loading: false,
};

beforeEach(() => {
  wrapper = render(<FoodCard {...mockProps} />);
});

describe('FoodCard Component', () => {
  it('should render component correctly', () => {
    const { getByTestId } = wrapper;
    expect(getByTestId('food-card')).toBeInTheDocument();
  });

  it('should call onClick when clicked', () => {
    const { getByTestId } = wrapper;
    const foodCard = getByTestId('food-card');

    fireEvent.click(foodCard);

    expect(mockProps.onClick).toHaveBeenCalled();
  });

  it('should match with snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
