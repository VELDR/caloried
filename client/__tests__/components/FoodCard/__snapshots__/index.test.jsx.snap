// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FoodCard Component should match with snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        data-testid="food-card"
      >
        <div>
          <div>
            <img
              alt="Test Food"
              src="http://localhost:8080test-image.jpg"
            />
            <span
              class="undefined undefined"
            >
              common
            </span>
          </div>
          <div>
            <div
              aria-label="Test Food"
              class=""
              data-mui-internal-clone-element="true"
            >
              Test Food
            </div>
            <div>
              <span
                aria-label="1 serving"
                class=""
                data-mui-internal-clone-element="true"
              >
                1
                 
                serving
              </span>
              <span>
                • 
                100
                 g
              </span>
            </div>
            <div>
              <div>
                <img
                  alt=""
                />
                200
              </div>
              <div>
                <img
                  alt=""
                  src="path/to/fat.png"
                />
                10
              </div>
              <div>
                <img
                  alt=""
                  src="path/to/fat.png"
                />
                20
              </div>
              <div>
                <img
                  alt=""
                  src="path/to/fat.png"
                />
                5
              </div>
            </div>
          </div>
        </div>
        <div>
          <div>
            <div
              aria-label="Protein"
              class=""
              data-mui-internal-clone-element="true"
            >
              <img
                alt=""
                src="path/to/fat.png"
              />
              10
               g
            </div>
            <div
              aria-label="Karbohidrat"
              class=""
              data-mui-internal-clone-element="true"
            >
              <img
                alt=""
                src="path/to/fat.png"
              />
              20
               g
            </div>
            <div
              aria-label="Lemak"
              class=""
              data-mui-internal-clone-element="true"
            >
              <img
                alt=""
                src="path/to/fat.png"
              />
              5
               g
            </div>
          </div>
          <div>
            <div>
              <span>
                200
              </span>
               kcal
            </div>
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ChevronRightIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 6 8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      data-testid="food-card"
    >
      <div>
        <div>
          <img
            alt="Test Food"
            src="http://localhost:8080test-image.jpg"
          />
          <span
            class="undefined undefined"
          >
            common
          </span>
        </div>
        <div>
          <div
            aria-label="Test Food"
            class=""
            data-mui-internal-clone-element="true"
          >
            Test Food
          </div>
          <div>
            <span
              aria-label="1 serving"
              class=""
              data-mui-internal-clone-element="true"
            >
              1
               
              serving
            </span>
            <span>
              • 
              100
               g
            </span>
          </div>
          <div>
            <div>
              <img
                alt=""
              />
              200
            </div>
            <div>
              <img
                alt=""
                src="path/to/fat.png"
              />
              10
            </div>
            <div>
              <img
                alt=""
                src="path/to/fat.png"
              />
              20
            </div>
            <div>
              <img
                alt=""
                src="path/to/fat.png"
              />
              5
            </div>
          </div>
        </div>
      </div>
      <div>
        <div>
          <div
            aria-label="Protein"
            class=""
            data-mui-internal-clone-element="true"
          >
            <img
              alt=""
              src="path/to/fat.png"
            />
            10
             g
          </div>
          <div
            aria-label="Karbohidrat"
            class=""
            data-mui-internal-clone-element="true"
          >
            <img
              alt=""
              src="path/to/fat.png"
            />
            20
             g
          </div>
          <div
            aria-label="Lemak"
            class=""
            data-mui-internal-clone-element="true"
          >
            <img
              alt=""
              src="path/to/fat.png"
            />
            5
             g
          </div>
        </div>
        <div>
          <div>
            <span>
              200
            </span>
             kcal
          </div>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="ChevronRightIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 6 8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
